include/boost/math/tools/polynomial.hpp

```diff
@@ -387,30 +387,37 @@ class polynomial :
    }
 
    template <class U>
-   polynomial& operator %=(const U& /*value*/)
+   polynomial& operator %=(const U& value)
    {
-       // We can always divide by a scalar, so there is no remainder:
-       *this = zero_element(std::multiplies<polynomial>());
+       // In the case that T is integral, this preserves the semantics
+       // p == r*(p/r) + (p % r), for polynomial<T> p and U r.
+       if (std::numeric_limits<T>::is_integer)
```

Comment:
I seem to recall that using `enable_if` based only on `T`, which is not one of the template parameters for the function, doesn't work. Let me check.